lib/Optimizer/CodeGen/ConvertToQIR.cpp

```diff
@@ -141,30 +138,26 @@ class QmemRAIIOpRewrite
     // Inspect the element type of the complex data, need to
     // know if its f32 or f64
     StringRef functionName;
-    if (Type eleTy = dyn_cast<LLVM::LLVMPointerType>(ccState.getType())
-                         .getElementType()) {
-      if (auto elePtrTy = dyn_cast<LLVM::LLVMPointerType>(eleTy))
-        eleTy = elePtrTy.getElementType();
-      if (auto arrayTy = dyn_cast<LLVM::LLVMArrayType>(eleTy))
-        eleTy = arrayTy.getElementType();
-      bool fromComplex = false;
-      if (auto complexTy = dyn_cast<LLVM::LLVMStructType>(eleTy)) {
-        fromComplex = true;
-        eleTy = complexTy.getBody()[0];
-      }
-      if (eleTy == rewriter.getI8Type())
-        functionName = cudaq::opt::QIRArrayQubitAllocateArrayWithCudaqStatePtr;
-      if (eleTy == rewriter.getF64Type())
-        functionName =
-            fromComplex
-                ? cudaq::opt::QIRArrayQubitAllocateArrayWithStateComplex64
-                : cudaq::opt::QIRArrayQubitAllocateArrayWithStateFP64;
-      if (eleTy == rewriter.getF32Type())
-        functionName =
-            fromComplex
-                ? cudaq::opt::QIRArrayQubitAllocateArrayWithStateComplex32
-                : cudaq::opt::QIRArrayQubitAllocateArrayWithStateFP32;
+    Type eleTy = raii.getInitElementType();
+    if (auto elePtrTy = dyn_cast<cudaq::cc::PointerType>(eleTy))
+      eleTy = elePtrTy.getElementType();
+    if (auto arrayTy = dyn_cast<cudaq::cc::ArrayType>(eleTy))
+      eleTy = arrayTy.getElementType();
+    bool fromComplex = false;
+    if (auto complexTy = dyn_cast<ComplexType>(eleTy)) {
+      fromComplex = true;
+      eleTy = complexTy.getElementType();
     }
+    if (isa<cudaq::cc::StateType>(eleTy))
```

Comment:
I added a conversion from `cc::StateType` to a `ptr<i8>` - does it need to be removed for this to work?

Reply:
(chatted offline)